File: python/game.py
from __future__ import absolute_import

from os import name
from os import system

from tictactoe.core import Player
from tictactoe.core import Slot
from tictactoe.core import Board


player_1 = Player('O', 'Player 1')
player_2 = Player('X', 'Player 2')


def clear():
    """
    Clear the screen.
    """
    clear_key = ['clear', 'cls'][name == 'nt']
    _ = system(clear_key)


def toggle_player(player):
    """
    Switch the current player instance.
    """
    return [player_1, player_2][player.name == player_1.name]


def main():
    """
    The main game logic.
    """
    game_board = Board(3, 3)
    current_player = player_1
    alert_message = ''
    win = False
    draw = False
    again = True

    while (not win) and (not draw) and again:
        clear()

        print('Guide\n')
        print(game_board.draw(use_tokens=False))
        print('\n\n')

        print('Playing Field\n')
        print(game_board.draw(use_tokens=True))
        print(alert_message)
        print(f'Turn of {current_player.name} [{current_player.token}]\n')

        move = input('Input: ')

        valid_move = game_board.write_turn(current_player, move)
        win = game_board.check_win()
        if valid_move:
            if not win:
                current_player = toggle_player(current_player)
            alert_message = ''
        else:
            alert_message = (
                f'Slot {str(move)} is either taken or invalid.\n\n')

        if (not win) and game_board.full:
            draw = True

        if win or draw:
            clear()

            print(game_board.draw(use_tokens=True))
            if win:
                print(f'{current_player.name} has won!\n\n')
            if draw:
                print('Draw!\n\n')

            play_again = input('Play again? (Y/N) : ')

            if any([play_again == answer for answer in ['Y', 'y']]):
                win = False
                again = True
                draw = False
                game_board.reset()
            else:
                win = False
                again = False
                draw = False


if __name__ == '__main__':
    main()

File: javascript/game.js
const { Player, Board } = require('./core/tictactoe');
const readline = require('readline');

const player_1 = new Player('O', 'Player 1');
const player_2 = new Player('X', 'Player 2');
const gameBoard = new Board(3, 3);
let currentPlayer = player_1;
let win = false;
let draw = false;
let again = true;
let validMove = false;
let alertMessage = '';

/**
 * Toggles player 1 and 2.
 * @param {Player} player - The current player.
 */
const togglePlayer = player => player.name == player_2.name ? player_1 : player_2;

/**
 * Returns a promise used for accepting inputs.
 * @param {string} text - The prompt message.
 * @param {Function} action
 * - The callback function used for processing the resolving value as an answer.
 * @param {boolean} returnInput
 * - Defaults as false. If true, the input is resolved back with the resolving value as an object.
 */
const input = (text, action, returnInput = false) => new Promise(resolve =>
{
  const interface = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  interface.question(text, _input =>
  {
    const answer = action(_input);
    interface.close();
    resolve(returnInput ? {answer, _input} : answer);
  });
});

/** The main game logic. */
async function main()
{
  while(!win && !draw && again)
  {
    console.clear();
    console.log('Guide\n');
    console.log(gameBoard.draw(false));
    console.log('\n\n');

    console.log('Playing Field\n');
    console.log(gameBoard.draw(true));
    console.log(alertMessage);
    console.log(`Turn of ${currentPlayer.name} [${currentPlayer.token}]\n`);

    const {answer: valid, _input: move} = await input(
      'Input: ',
      input => gameBoard.writeTurn(currentPlayer, input),
      true
    );
    validMove = valid;
    win = gameBoard.checkWin();

    if(validMove)
    {
      if(!win)
      {
        currentPlayer = togglePlayer(currentPlayer);
      }
      alertMessage = '';
    }
    else
    {
      alertMessage = `Slot ${move} is either taken or invalid.\n\n`;
    }

    if(!win && gameBoard.isFull())
    {
      draw = true;
    }

    if(win || draw)
    {
      console.clear();
      console.log(gameBoard.draw(true));
      console.log(win ? `${currentPlayer.name} has won!` : 'Draw!');
      console.log('\n\n');

      await input('Play again? (Y/N): ', input => {
        if(['Y', 'y'].includes(input))
        {
          win = false;
          again = true;
          draw = false;
          gameBoard.reset();
        }
        else
        {
          win = false;
          again = false;
          draw = false;
        }
      });
    }
  }
}

/* Export the main function for potential import. */
module.exports = main;

main();
